### PR TITLE
MULE-9334: DB Connection leak with an invalid table in Select statement

### PR DESCRIPTION
**Pull request by elrodro83**

### CODE REVIEW

File: distributions/standalone/src/main/resources/MIGRATION.txt [hunk 1]
```diff
@@ -5,6 +5,7 @@ The following is a list of changes you have to execute to migrate from Mule Runt
 Please check https://developer.mulesoft.com/docs/display/current/Mule+ESB for more details.
 
 Migration changes from 3.7.x to 3.8.x
+MULE-9334: When an SQLException occurs while executing a query, all ResultSets from that db connection (i.e.: multiple queries in the same transaction) that remain open will be closed.
```
Comment (pablokraan) [inline distributions/standalone/src/main/resources/MIGRATION.txt]: db > DB

File: modules/db/src/main/java/org/mule/module/db/internal/config/processor/AbstractAdvancedDbProcessorDefinitionParser.java [hunk 2]
```diff
@@ -104,4 +108,8 @@ protected void parseMetadataProvider(Element element, BeanDefinitionBuilder buil
         }
     }
 
+    protected void configureStatementResultSetCloser(BeanDefinitionBuilder builder)
+    {
+        builder.addPropertyReference(STATEMENT_STREAMING_RESULTSET_CLOSER_PROPERTY, "db.statementStreamingResultSetCloser");
```
Comment (pablokraan) [inline modules/db/src/main/java/org/mule/module/db/internal/config/processor/AbstractAdvancedDbProcessorDefinitionParser.java]: I would add "mule."  prefix

File: modules/db/src/main/java/org/mule/module/db/internal/result/statement/StatementStreamingResultSetCloser.java [hunk 3]
```diff
@@ -20,11 +20,36 @@
...
     private final ConcurrentHashMap<DbConnection, Set<ResultSet>> connectionResultSets = new ConcurrentHashMap<DbConnection, Set<ResultSet>>();
 
     private final ConcurrentHashMap<DbConnection, Object> connectionLocks = new ConcurrentHashMap<DbConnection, Object>();
 
+    /**
+     * Closes all tracked {@link ResultSet}s for the passed {@code connection}.
+     * 
+     * @param connection
+     */
+    public void closeAll(DbConnection connection)
```
Comment (pablokraan) [inline modules/db/src/main/java/org/mule/module/db/internal/result/statement/StatementStreamingResultSetCloser.java]: closeAll > closeResultsets?

File: modules/db/src/main/java/org/mule/module/db/internal/result/statement/StatementStreamingResultSetCloser.java [hunk 4]
```diff
@@ -20,11 +20,36 @@
...
+        synchronized (connectionLock)
+        {
+            checkValidConnectionLock(connection, connectionLock);
+
+            Set<ResultSet> resultSets = connectionResultSets.get(connection);
+            if (resultSets != null)
+            {
+                for (ResultSet resultSet : resultSets)
+                {
+                    super.close(connection, resultSet);
```
Comment (pablokraan) [inline modules/db/src/main/java/org/mule/module/db/internal/result/statement/StatementStreamingResultSetCloser.java]: Should not this super.close be wrapped in a try/catch?

File: modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    public SelectStreamingExceptionTestCase(String dataSourceConfigResource, AbstractTestDatabase testDatabase)
+    {
+        super(dataSourceConfigResource, testDatabase);
+    }
+
+    @Parameterized.Parameters
+    public static List<Object[]> parameters()
+    {
+        return Collections.singletonList(new Object[] {"integration/config/derby-pooling-db-config.xml", new DerbyTestDatabase()});
```
Comment (pablokraan) [inline modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java]: Do you need a config with pooling?

File: modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        return config.getDataSource();
+    }
+
+    @Test
+    public void streamingException() throws Exception
+    {
+        LocalMuleClient client = muleContext.getClient();
+
+        // Assuming the pool has a max connections of 2
+        for (int i = 0; i < 3; ++i)
```
Comment (pablokraan) [inline modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java]: Use a constant for 3, like MAX_CONNECTIONS and remove the comment

File: modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            MuleMessage response = client.send("vm://selectStreamingException", new DefaultMuleMessage(TEST_MESSAGE, muleContext));
```
Comment (pablokraan) [inline modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java]: Try to write the test using runFlow

File: modules/db/src/main/java/org/mule/module/db/internal/config/processor/AbstractAdvancedDbProcessorDefinitionParser.java (hunk 2 above)
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/config/processor/AbstractAdvancedDbProcessorDefinitionParser.java]: Let's keep the db. prefix to be consistent with the other property names.

File: modules/db/src/main/java/org/mule/module/db/internal/result/statement/StatementStreamingResultSetCloser.java (hunk 4 above)
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/result/statement/StatementStreamingResultSetCloser.java]: The called method is already doing a try/catch.

File: modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java (hunk 5 above)
Comment (elrodro83) [inline modules/db/src/test/java/org/mule/module/db/integration/select/SelectStreamingExceptionTestCase.java]: In this case i need a pool so i have a certain number of iterations to run before the test fails.

File: modules/db/src/test/resources/integration/select/select-streaming-exception-config.xml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    <flow name="selectStreamingException">
+        <inbound-endpoint address="vm://selectStreamingException" exchange-pattern="request-response"/>
+
+        <db:select config-ref="pooledJdbcConfig" streaming="true">
+            <db:parameterized-query>select * from NOT_EXISTS order by ID</db:parameterized-query>
+        </db:select>
+    </flow>
+
+    <flow name="selectExceptionClosesPreviousResultSets">
+        <inbound-endpoint address="vm://selectExceptionClosesPreviousResultSets" exchange-pattern="request-response"/>
```
Comment (pablokraan) [inline modules/db/src/test/resources/integration/select/select-streaming-exception-config.xml]: Remove inbound endpoints

**Pull request merged**